be/src/runtime/memory/counting_allocator.h

```diff
@@ -0,0 +1,81 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <iostream>
+#include <memory>
+
+#include "exprs/expr_context.h"
+
+namespace starrocks {
+
+template <class T>
+class CountingAllocator {
+public:
+    typedef T value_type;
+    typedef size_t size_type;
+    using propagate_on_container_copy_assignment = std::true_type; // for consistency
+    using propagate_on_container_move_assignment = std::true_type; // to avoid the pessimization
+    using propagate_on_container_swap = std::true_type;            // to avoid the undefined behavior
+
+    template <typename U>
+    struct rebind {
+        using other = CountingAllocator<U>;
+    };
+    CountingAllocator() = default;
```

Comment:
CountingAllocator() = delete? otherwise T* allocate(size_t n) maybe crash?

Reply:
> CountingAllocator() = delete? otherwise T* allocate(size_t n) maybe crash?

we can't delete this constructor because it may be used by phmap, as a default parameter value.

we need to ensure that the _counter is initialized in allocator when creating the phmap.

I have add a DCHECK in allocate and deallocate function

```cpp
    explicit raw_hash_set(size_t bucket_cnt, const hasher& hashfn = hasher(), const key_equal& eq = key_equal(),
                          const allocator_type& alloc = allocator_type())
            : ctrl_(EmptyGroup()), settings_(0, hashfn, eq, alloc) {
        if (bucket_cnt) {
            capacity_ = NormalizeCapacity(bucket_cnt);
            reset_growth_left();
            initialize_slots();
        }
    }

```